AWS: Fix KeyError in aws_secret lookup

bobdoah opened the pull request:
The region parameter must be specified as either a keyword arg
or in the lookup plugin options. I opted for the latter, as
`run` method currently retrieves it from the options.

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
I added the `region` to the options docstring. This stops the lookup plugin from blowing up as it's not specified. It's optional, so does not have to be supplied - but the plugin was raising a KeyError because it was completely undefined. The plugin was already written to read the `region` value from the options dictionary. The other approach is to supply it as a keyword arg, as in the `lib/ansible/plugins/lookup/aws_ssm.py`. 

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #54790 

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
lib/ansible/plugins/lookup/aws_secret.py

##### ADDITIONAL INFORMATION

bobdoah committed:
Fixes #54790

The region parameter must be specified as either a keyword arg
or in the lookup plugin options. I opted for the latter, as
`run` method currently retrieves it from the options.

Akasurde commented on lib/ansible/plugins/lookup/aws_secret.py:
```diff
@@ -36,6 +36,9 @@
     type: boolean
     default: false
+  region:
+    description: AWS region in which to do the lookup
```
```suggestion
    description: AWS region in which to do the lookup.
```

Akasurde reviewed (changes requested):
Would you be able to add unit test or integration for this ? Thanks.

Akasurde committed:
Update lib/ansible/plugins/lookup/aws_secret.py

Co-Authored-By: bobdoah <robertwilliams1985@gmail.com>

bobdoah committed:
Added unit tests as requested by @Akasurde.

The tests consist of a simple success and failure case. I noticed
that the AWS SSM lookup used an include of the aws_region docs to ensure
the config option was set correctly. I've swapped to this, instead of
defining it manually. It seems to work fine with the use case I
specified in #54790.

Whilst I was testing I dumped a pytestdebug.log, which git wanted to
add. I've added it to the .gitignore file to stop anyone else
accidentally committing this.

bobdoah commented:
> Would you be able to add unit test or integration for this ? Thanks.

@Akasurde, I've added a couple of unit tests which should cover this.

bobdoah committed:
Fixed pep8 non-compliance.

bobdoah committed:
The tzlocal package isn't installed by default. This is needed to mock the CreatedDate in the AWS response correctly.

bobdoah commented:
ready_for_review

mattclay commented on test/units/plugins/lookup/test_aws_secret.py:
```diff
@@ -0,0 +1,88 @@
+
+import pytest
+import datetime
+import tzlocal
```
Instead of adding a dependency on `tzlocal`, use UTC for the timezone. Something like this should work:

```python
class UTC(datetime.tzinfo):
    ZERO = datetime.timedelta(0)

    def utcoffset(self, dt):
        return self.ZERO

    def tzname(self, dt):
        return "UTC"

    def dst(self, dt):
        return self.ZERO


utc = UTC()

value = datetime.datetime(2019, 4, 4, 11, 41, 0, 878000, tzinfo=utc)
```

See the "Example tzinfo classes" portion of the [tzinfo documentation](https://docs.python.org/2/library/datetime.html#tzinfo-objects) for the example on which this is based.

bobdoah replied:
Is there a reason not to add the dependency? The `tzlocal` library is just a wrapper around `pytz` and the [python docs](https://docs.python.org/2/library/datetime.html#tzinfo-objects) say

> The standard library has no tzinfo instances, but there exists a third-party library which brings the IANA timezone database (also known as the Olson database) to Python: pytz.
>
> pytz contains up-to-date information and its usage is recommended.

mattclay replied:
After looking at this more closely I realized that boto3 does use local time here instead of UTC. However, boto3 is using `dateutil` instead of `tzlocal`. So this will work without any additional requirements:

```python
import dateutil.tz

value = datetime.datetime(2019, 4, 4, 11, 41, 0, 878000, tzinfo=dateutil.tz.tz.tzlocal())
```

bobdoah committed:
Remove tzlocal and rely on dateutils built in timezone module, as suggested by mattclay

bobdoah commented:
@mattclay  @Akasurde is there something else I need to do to get this merged?

AndrewCi commented:
Any updates on this?

mattclay commented on .gitignore:
```diff
@@ -65,6 +65,7 @@ coverage.xml
 /test/units/cover-html
 /test/integration/targets/*/backup/
 /test/cache/*
+pytestdebug.log
```
Leave this out of the PR. If it's something you'd like to include, open a separate PR with an explanation of why it's needed.

bobdoah replied:
Argh, I tried to remove this from history and made a mess.

bobdoah replied:
Phew, thankfully ORIG_HEAD came to the rescue.

mattclay commented on test/units/plugins/lookup/test_aws_secret.py:
```diff
@@ -0,0 +1,88 @@
+
+import pytest
+import datetime
+import dateutil.tz
```
This should have a try/except around it as well, since `dateutil.tz` isn't in the standard library.

mattclay commented on test/units/plugins/lookup/test_aws_secret.py:
```diff
@@ -0,0 +1,88 @@
+except ImportError:
+    pytestmark = pytest.mark.skip("This test requires the boto3 and botocore Python libraries")
+
+simple_variable_success_response = {
```
Put this in the test function that uses it, since it depends on `dateutil.tz`. If it needs to be shared across multiple test functions it can be put in a fixture instead.

DenysTT commented:
Any idea when it will be released or how this issue can be avoided?
Thanks

StubbsPKS commented:
Is there a workaround for this issue until the merge is complete and released?

bobdoah committed:
The relevant test should be skipped if dateutil.tz is not importable.

I used pytests importorskip method to make sure that the success case
will be skipped if dateutil.tz library isn't available for the mock data
definition. Uninstalling it actually causes the boto import to fail, and
the entire module to be skipped as it stands. That might not always be
the case.

I've also moved the shared mock credentials into a fixture included by
both tests rather than the module-level variable. This parallels what
@mattclay requested for the simple_variable_success_response, however
I moved that inside the test as it's only required by a single test.

Removed the gitignore changes from the PR as @mattclay asks.

bobdoah commented:
@mattclay I believe I've resolved the points you raised

bobdoah commented:
> @mattclay I believe I've resolved the points you raised

or I would've done if I'd not forgotten to `git add`

bobdoah commented:
Does this need @Akasurde to approve the changes too?

bkmeneguello commented:
@Akasurde  I've tested and this works perfectly

yourbuddyconner commented:
Would love for this to be merged right about now.

hibbert commented:
What ansible version will the fix be in? I'm on 2.8.1 and still get the error

geekofalltrades commented:
Was this supposed to have made 2.8.2? It's still broken for me.

https://github.com/ansible/ansible/issues/54790#issuecomment-508283664

bartlabno commented:
https://github.com/ansible/ansible/blob/v2.8.2/lib/ansible/plugins/lookup/aws_secret.py

aws_secret doesn't contain changes from this fix I am afraid @bobdoah

s-hertel commented:
This was merged for 2.9. I opened a backport for the next minor 2.8 release - #58722. If a patch you need has not been backported there is also a process documented here https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs to take things into your own hands. Asking in IRC may also get a quicker response as closed tickets aren't really monitored.
